### PR TITLE
[OF-1107] feat: ECommerceSystem and GetProductInformation

### PR DESCRIPTION
**Pull request by Mag-JB**
connected-spaces-platform Pull Request

**For the reviewer**

* [ ] If required, are the changes covered by appropriate tests?
* [ ] Are any public-facing API changes well documented?
* [ ] Is the code easily readable and extensible and/or follow existing conventions?

### CODE REVIEW

**Comment by MAG-AdamThorn** [comment]
Are you planning to document API in a subsequent commit?

**Comment by Mag-JB** [comment]
> Are you planning to document API in a subsequent commit?

Yes, this is just to unblock other Ecommerce tickets

File: Library/include/CSP/Systems/ECommerce/ECommerce.h [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+public:
+	ProductInfo(const csp::common::String& IdIn,
+				const csp::common::String& TitleIn,
+				bool AvailableForSaleIn,
+				const ProductImageInfo& ImageIn,
+				const csp::common::Map<csp::common::String, csp::common::String>& SelectedOptionsIn,
+				double UnitPriceIn,
+				const ProductMediaInfo& MediaIn);
+
+	csp::common::String Id;
```
Comment (MAG-SamBirley) [inline Library/include/CSP/Systems/ECommerce/ECommerce.h:77]: Can we get inline doxygen-format comments explaining what each of these does?

File: Library/include/CSP/Systems/ECommerce/ECommerce.h [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+};
+
+/// @brief Represents image information for a Product
+class CSP_API ProductImageInfo
+{
+public:
+	ProductImageInfo() = default;
+	ProductImageInfo(const csp::common::String& AltIn, const csp::common::String& UrlIn, const int64_t WidthIn, const int64_t HeightIn);
+
+	csp::common::String Alt;
```
Comment (MAG-SamBirley) [inline Library/include/CSP/Systems/ECommerce/ECommerce.h:59]: Can we get inline doxygen-format comments explaining what each of these does?

File: Library/include/CSP/Systems/ECommerce/ECommerce.h [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+{
+public:
+	ProductMediaInfo() = default;
+	ProductMediaInfo(const csp::common::String& MediaContentTypeIn,
+					 const csp::common::String& AltIn,
+					 const csp::common::String& UrlIn,
+					 const int64_t WidthIn,
+					 const int64_t HeightIn);
+
+	csp::common::String MediaContentType;
```
Comment (MAG-SamBirley) [inline Library/include/CSP/Systems/ECommerce/ECommerce.h:45]: Can we get inline doxygen-format comments explaining what each of these does?

File: Library/include/CSP/Systems/ECommerce/ECommerceSystem.h [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+
+class WebClient;
+
+}
+
+
+namespace csp::systems
+{
+/// @ingroup eCommerce System
+/// @brief Public facing system that allows interfacing with CSP's concept of a eCommerce platform.
```
Comment (MAG-SamBirley) [inline Library/include/CSP/Systems/ECommerce/ECommerceSystem.h:39]: 'an e-commerce platform'

File: Library/include/CSP/Systems/ECommerce/ECommerceSystem.h [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+{
+
+class WebClient;
+
+}
+
+
+namespace csp::systems
+{
+/// @ingroup eCommerce System
```
Comment (MAG-SamBirley) [inline Library/include/CSP/Systems/ECommerce/ECommerceSystem.h:38]: Uppercase E here?

File: Library/src/Systems/ECommerce/ECommerce.cpp [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+ */
+#include "CSP/Systems/eCommerce/ECommerce.h"
+
+
+namespace csp::systems
+{
+
+ProductImageInfo::ProductImageInfo(const csp::common::String& AltIn, const csp::common::String& UrlIn, const int64_t WidthIn, const int64_t HeightIn)
+{
+	Alt	   = AltIn;
```
Comment (MAG-SamBirley) [inline Library/src/Systems/ECommerce/ECommerce.cpp:24]: Prefer member initialisations here?
```
MyCtor(ThingA)
: MemberA(ThingA)
{}
```
style?

**Review by MAG-SamBirley — CHANGES_REQUESTED** [review]

File: Library/src/Systems/ECommerce/ECommerce.cpp (hunk 6 above)
Comment (Mag-JB) [reply Library/src/Systems/ECommerce/ECommerce.cpp:24]: sure

Comment (Mag-JB) [reply Library/src/Systems/ECommerce/ECommerce.cpp:24]: sorted

File: Library/include/CSP/Systems/ECommerce/ECommerceSystem.h (hunk 5 above)
Comment (Mag-JB) [reply Library/include/CSP/Systems/ECommerce/ECommerceSystem.h]: sorted

File: Library/include/CSP/Systems/ECommerce/ECommerceSystem.h (hunk 4 above)
Comment (Mag-JB) [reply Library/include/CSP/Systems/ECommerce/ECommerceSystem.h]: went with ECommerce

File: Library/include/CSP/Systems/ECommerce/ECommerce.h (hunk 1 above)
Comment (Mag-JB) [reply Library/include/CSP/Systems/ECommerce/ECommerce.h:81]: should be good

File: Library/include/CSP/Systems/ECommerce/ECommerce.h (hunk 2 above)
Comment (Mag-JB) [reply Library/include/CSP/Systems/ECommerce/ECommerce.h:62]: should be good

File: Library/include/CSP/Systems/ECommerce/ECommerce.h (hunk 3 above)
Comment (Mag-JB) [reply Library/include/CSP/Systems/ECommerce/ECommerce.h:47]: Should be good

**Review by MAG-SamBirley — DISMISSED** [review]

File: Library/include/CSP/Systems/ECommerce/ECommerce.h [hunk 7]
```diff
@@ -0,0 +1,155 @@
...
+/// @brief Represents variant information for a product
+class CSP_API ProductVariantInfo
+{
+public:
+	ProductVariantInfo() = default;
+	/// @brief Id of the variant.
+	csp::common::String Id;
+	/// @brief Title of the variant.
+	csp::common::String Title;
+	/// @brief Title of variant.
```
Comment (MAG-AdamThorn) [inline Library/include/CSP/Systems/ECommerce/ECommerce.h:90]: You have a duplicate of 'Title of the variant'

File: Library/include/CSP/Systems/ECommerce/ECommerce.h [hunk 8]
```diff
@@ -0,0 +1,155 @@
...
+	csp::common::String Url;
+	/// @brief is variant available for sale.
```
Comment (MAG-AdamThorn) [inline Library/include/CSP/Systems/ECommerce/ECommerce.h:92]: Capitalise the 'i' - Is variant available...

File: Library/include/CSP/Systems/ECommerce/ECommerce.h [hunk 9]
```diff
@@ -0,0 +1,155 @@
...
+	bool AvailableForSale;
+	/// @brief This image for a variant
+	ProductMediaInfo Media;
```
Comment (MAG-AdamThorn) [inline Library/include/CSP/Systems/ECommerce/ECommerce.h:95]: Is MediaInfo always an image? Not clear from documentation of ProductMediaInfo above.

File: Library/include/CSP/Systems/ECommerce/ECommerce.h [hunk 10]
```diff
@@ -0,0 +1,155 @@
...
+class CSP_API ProductInfo
+{
+public:
+	ProductInfo() = default;
+
+	/// @brief Id of the product.
+	csp::common::String Id;
+	/// @brief Title of the product.
+	csp::common::String Title;
+	/// @brief  Time the product was created.
```
Comment (MAG-AdamThorn) [inline Library/include/CSP/Systems/ECommerce/ECommerce.h:114]: Extra whitespace between '@brief and Time'.

File: Library/include/CSP/Systems/ECommerce/ECommerce.h [hunk 11]
```diff
@@ -0,0 +1,155 @@
...
+	csp::common::String CreatedAt;
+	/// @brief Description of the product.
+	csp::common::String Description;
+	/// @brief array of product variants.
+	csp::common::Array<ProductVariantInfo> Variants;
+	/// @brief array of product tags.
+	csp::common::Array<csp::common::String> Tags;
+	/// @brief This array holds  media for the product
```
Comment (MAG-AdamThorn) [inline Library/include/CSP/Systems/ECommerce/ECommerce.h:122]: Additional whitespace between 'holds' and 'media'

File: Library/include/CSP/Systems/ECommerce/ECommerceSystem.h [hunk 12]
```diff
@@ -0,0 +1,61 @@
...
+
+}
+
+
+namespace csp::systems
+{
+/// @ingroup ECommerce System
+/// @brief Public facing system that allows interfacing with CSP's concept of a ECommerce platform.
+/// Offers methods for utilising shopify through CSP
+class CSP_API CSP_NO_DISPOSE ECommerceSystem : public SystemBase
```
Comment (MAG-AdamThorn) [inline Library/include/CSP/Systems/ECommerce/ECommerceSystem.h:41]: Do we want to refer to 'shopify' in this class? Would we have to remove these references in the future when we support other platforms, or is that not a concern for now?

File: Library/include/CSP/Systems/ECommerce/ECommerce.h (hunk 11 above)
Comment (Mag-JB) [reply Library/include/CSP/Systems/ECommerce/ECommerce.h]: sorted

File: Library/include/CSP/Systems/ECommerce/ECommerceSystem.h (hunk 12 above)
Comment (Mag-JB) [reply Library/include/CSP/Systems/ECommerce/ECommerceSystem.h:41]: sorted

File: Library/include/CSP/Systems/ECommerce/ECommerce.h (hunk 7 above)
Comment (Mag-JB) [reply Library/include/CSP/Systems/ECommerce/ECommerce.h]: sorted

File: Library/include/CSP/Systems/ECommerce/ECommerce.h (hunk 9 above)
Comment (Mag-JB) [reply Library/include/CSP/Systems/ECommerce/ECommerce.h:95]: sorted

File: Library/include/CSP/Systems/ECommerce/ECommerce.h [hunk 13]
```diff
@@ -0,0 +1,155 @@
...
+
+	/// @brief Id of the product.
+	csp::common::String Id;
+	/// @brief Title of the product.
+	csp::common::String Title;
+	/// @brief  Time the product was created.
+	csp::common::String CreatedAt;
+	/// @brief Description of the product.
+	csp::common::String Description;
+	/// @brief array of product variants.
```
[the remark this answers is not in the document]
Comment (Mag-JB) [reply Library/include/CSP/Systems/ECommerce/ECommerce.h]: sorted

File: Library/include/CSP/Systems/ECommerce/ECommerce.h (hunk 10 above)
Comment (Mag-JB) [reply Library/include/CSP/Systems/ECommerce/ECommerce.h]: sorted

File: Tests/CSharp/src/Tests/ECommerceSystemTests.cs [hunk 14]
```diff
@@ -0,0 +1,118 @@
...
+            VariantIds[0] = "gid://shopify/ProductVariant/46314311516433";
+            VariantIds[1] = "gid://shopify/ProductVariant/46314311647505";
+            VariantIds[2] = "gid://shopify/ProductVariant/46314311745809";
+            VariantIds[3] = "gid://shopify/ProductVariant/46314311844113";
+            // Log in
+            _ = UserSystemTests.LogIn(userSystem);
+            
+            var details = getShopifyDetails();
+
+            using var result = eCommerceSystem.GetProductInformation(details["SpaceId"], details["ProductId"]).Result;
```
Comment (MAG-AdamThorn) [inline Tests/CSharp/src/Tests/ECommerceSystemTests.cs:70]: I must be missing something, but where in this test are you creating the Product that you are retrieving in the GetProductInformation() method above?

File: Library/include/CSP/Systems/ECommerce/ECommerceSystem.h (hunk 12 above)
Comment (MAG-AdamThorn) [reply Library/include/CSP/Systems/ECommerce/ECommerceSystem.h:41]: Still refer to Shopify below: csp::services::ApiBase* ShopifyAPI;

File: Tests/CSharp/src/Tests/ECommerceSystemTests.cs (hunk 14 above)
Comment (Mag-JB) [reply Tests/CSharp/src/Tests/ECommerceSystemTests.cs:70]: These are manual tests as there no way to automate the creation of products at the moment. The test uses the quickstart store item Gift Card which should be identical to all uses when they set up their shopify account. these are turn off from TC tests.

File: Library/include/CSP/Systems/ECommerce/ECommerceSystem.h (hunk 12 above)
Comment (Mag-JB) [reply Library/include/CSP/Systems/ECommerce/ECommerceSystem.h:41]: I kept that as the CHS endpoint is ShopifyApi which we always refer to as ShopifyAPI same with GroupApi and GroupAPI

**Review by MAG-AdamThorn — APPROVED** [review]

**Review by MAG-mv — APPROVED** [review]

**Pull request merged**